File: src/com/eagle/programmar/Rust/Statements/Rust_Block_Statement.java
// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Aug 24, 2025

package com.eagle.programmar.Rust.Statements;

import java.util.ArrayList;

import com.eagle.interpret.EagleInterpreter;
import com.eagle.interpret.EagleRunnableWithResult;
import com.eagle.programmar.Rust.Rust_Generator;
import com.eagle.programmar.Rust.Rust_Statement;
import com.eagle.tokens.AbstractToken;
import com.eagle.tokens.TokenList;
import com.eagle.tokens.TokenSequence;
import com.eagle.tokens.interfaces.AbstractStatement;
import com.eagle.tokens.punctuation.PunctuationLeftBrace;
import com.eagle.tokens.punctuation.PunctuationRightBrace;
import com.eagle.transform.EagleGenerator;
import com.eagle.transform.EagleTransformableStatement;
import com.eagle.transform.EagleTransformer;

public class Rust_Block_Statement extends TokenSequence
		implements EagleRunnableWithResult, EagleTransformableStatement
{
	public @S(10) PunctuationLeftBrace leftBrace;
	public @S(20) TokenList<Rust_Statement> statements;
	public @S(30) PunctuationRightBrace rightBrace;

	@Override
	public Eagle_Statement_Result interpretStatement(EagleInterpreter interpreter)
	{
		Eagle_Statement_Result result = Eagle_Statement_Result.NORMAL;
		for (Rust_Statement stmt : statements._elements)
		{
			result = interpreter.tryToInterpret(stmt);
			if (result != Eagle_Statement_Result.NORMAL)
			{
				break;
			}
		}
		return result;
	}

	@Override
	public AbstractStatement transformStatement(EagleTransformer transformer, EagleGenerator generator)
	{
		ArrayList<AbstractStatement> result = new ArrayList<AbstractStatement>();
		for (Rust_Statement statement : statements._elements)
		{
			ArrayList<AbstractStatement> stmts = transformer.transformStatement(generator, statement.getWhich());
			if (stmts != null)
			{
				for (AbstractStatement stmt : stmts)
				{
					result.add(stmt);
				}
			}
		}

		return generator.newBlockStatement(result, this);
	}

	public Rust_Statement generateBlock(ArrayList<Rust_Statement> stmts,
			AbstractToken source)
	{
		this.leftBrace = new PunctuationLeftBrace();
		this.rightBrace = new PunctuationRightBrace();
		this.statements = new TokenList<Rust_Statement>();
		this.statements.setPresent(true);
		for (Rust_Statement stmt : stmts)
		{
			this.statements.addToken(stmt);
		}
		return Rust_Generator.wrapStatement(this);
	}
	
	public static ArrayList<AbstractStatement> collectStatements(EagleTransformer transformer,
			EagleGenerator generator, Rust_Statement statement)
	{
		// Lots of extra work here to avoid duplicated braces; {{stmts}} is not nice.
		ArrayList<AbstractStatement> newStmts;
		if (statement.getWhich() instanceof Rust_Block_Statement)
		{
			Rust_Block_Statement block = (Rust_Block_Statement) statement.getWhich();
			newStmts = new ArrayList<AbstractStatement>();
			for (Rust_Statement blockStmt : block.statements._elements)
			{
				ArrayList<AbstractStatement> oneStmt = transformer.transformStatement(generator, blockStmt.getWhich());
				if (oneStmt != null)
				{
					for (AbstractStatement newStmt : oneStmt)
					{
						newStmts.add(newStmt);
					}
				}
			}
		}
		else
		{
			// Rare case I think, def fn = stmt, with no braces
			newStmts = transformer.transformStatement(generator, statement.getWhich());
		}
		return newStmts;
	}
}

File: csrc/com/eagle/programmar/Rust/Statements/Rust_Block_Statement.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// Purchase a Premium Edition license at:
// https://www.tangiblesoftwaresolutions.com/order/order-java-to-csharp.html
// ====================================================================================================

using System.Collections.Generic;

// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Aug 24, 2025

namespace com.eagle.programmar.Rust.Statements
{

	using EagleInterpreter = com.eagle.interpret.EagleInterpreter;
	using EagleRunnableWithResult = com.eagle.interpret.EagleRunnableWithResult;
	using Rust_Generator = com.eagle.programmar.Rust.Rust_Generator;
	using Rust_Statement = com.eagle.programmar.Rust.Rust_Statement;
	using AbstractToken = com.eagle.tokens.AbstractToken;
	using TokenList = com.eagle.tokens.TokenList;
	using TokenSequence = com.eagle.tokens.TokenSequence;
	using AbstractExpression = com.eagle.tokens.interfaces.AbstractExpression;
	using AbstractStatement = com.eagle.tokens.interfaces.AbstractStatement;
	using AbstractType = com.eagle.tokens.interfaces.AbstractType;
	using AbstractVariable = com.eagle.tokens.interfaces.AbstractVariable;
	using PunctuationLeftBrace = com.eagle.tokens.punctuation.PunctuationLeftBrace;
	using PunctuationRightBrace = com.eagle.tokens.punctuation.PunctuationRightBrace;
	using EagleGenerator = com.eagle.transform.EagleGenerator;
	using EagleTransformableStatement = com.eagle.transform.EagleTransformableStatement;
	using EagleTransformer = com.eagle.transform.EagleTransformer;

	public class Rust_Block_Statement : TokenSequence, EagleRunnableWithResult, EagleTransformableStatement
	{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) @INDENT PunctuationLeftBrace leftBrace;
		public  INDENT;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.tokens.TokenList<com.eagle.programmar.Rust.Rust_Statement> statements;
		public TokenList<Rust_Statement> statements;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) @OUTDENT PunctuationRightBrace rightBrace;
		public  OUTDENT;

		public override Eagle_Statement_Result interpretStatement(EagleInterpreter interpreter)
		{
			Eagle_Statement_Result result = Eagle_Statement_Result.NORMAL;
			foreach (Rust_Statement stmt in statements._elements)
			{
				result = interpreter.tryToInterpret(stmt);
				if (result != Eagle_Statement_Result.NORMAL)
				{
					break;
				}
			}
			return result;
		}

		public override AbstractStatement transformStatement(EagleTransformer transformer, EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
		{
			List<AbstractStatement> result = new List<AbstractStatement>();
			foreach (Rust_Statement statement in statements._elements)
			{
				List<AbstractStatement> stmts = transformer.transformStatement(generator, statement.getWhich());
				if (stmts != null)
				{
					foreach (AbstractStatement stmt in stmts)
					{
						result.Add(stmt);
					}
				}
			}

			return generator.newBlockStatement(result, this);
		}

		public static Rust_Statement generateBlock(List<Rust_Statement> stmts, AbstractToken source)
		{
			Rust_Block_Statement block = new Rust_Block_Statement();
			block.leftBrace = new PunctuationLeftBrace();
			block.rightBrace = new PunctuationRightBrace();
			block.statements = new TokenList<Rust_Statement>();
			block.statements.setPresent(true);
			foreach (Rust_Statement stmt in stmts)
			{
				block.statements.addToken(stmt);
			}
			return Rust_Generator.wrapStatement(block);
		}

		public static List<AbstractStatement> collectStatements(EagleTransformer transformer, EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator, Rust_Statement statement)
		{
			// Lots of extra work here to avoid duplicated braces; {{stmts}} is not nice.
			List<AbstractStatement> newStmts;
			if (statement.getWhich() is Rust_Block_Statement)
			{
				Rust_Block_Statement block = (Rust_Block_Statement) statement.getWhich();
				newStmts = new List<AbstractStatement>();
				foreach (Rust_Statement blockStmt in block.statements._elements)
				{
					List<AbstractStatement> oneStmt = transformer.transformStatement(generator, blockStmt.getWhich());
					if (oneStmt != null)
					{
						foreach (AbstractStatement newStmt in oneStmt)
						{
							newStmts.Add(newStmt);
						}
					}
				}
			}
			else
			{
				// Rare case I think, def fn = stmt, with no braces
				newStmts = transformer.transformStatement(generator, statement.getWhich());
			}
			return newStmts;
		}
	}

}
